Improve position generation

sir-marc opened the pull request:
- Extracts position generation into function
- Makes sure an object is spawned on a free field

sir-marc committed:
Move GameObjects type into own folder

sir-marc committed:
Create Food and Unit interfaces

sir-marc committed:
Fix Food rendering

sir-marc committed:
Fix Unit props interface

sir-marc committed:
Add Food to gameObjects rendering

sir-marc committed:
Add count gameObjects helper

sir-marc committed:
Add food spawner

sir-marc committed:
Adapt MainController to new Unit Interface

sir-marc committed:
Add FoodSpawner into tick cycle

sir-marc committed:
Extract magic number into constant

sir-marc committed:
Integrate food-spawning into game-engine

sir-marc committed:
Integrate food-spawning into game-engine

sir-marc committed:
Refactor statebuilder test helper

sir-marc committed:
Add food spawning tests

sir-marc committed:
Add gameObjectType typesafety

sir-marc committed:
Add gameObjectType typesafety

sir-marc committed:
Fix broken imports

sir-marc committed:
Fix out of scope imports

sir-marc committed:
Revert "Fix broken imports"

This reverts commit 537e3417fa72b439a5c7b33821a23d51934a81d1.

sir-marc committed:
Remove immer from food-spawner to not make state immutable

For now at least

sir-marc committed:
Fix broken import statement

sir-marc committed:
Extract position generation

sir-marc committed:
Extract position generation

sir-marc committed:
Fix occupied fields check

StuckiSimon committed:
AB#46 Provide User Context to frontend (#43)

Render unit of current user differently.

StuckiSimon committed:
Remove Leftovers from frontend bootstrap (#44)

StuckiSimon committed:
Refactor Frontend Service Layer (#45)

michaelschufi committed:
AB#100 npm script cleanup (#46)

* clean up of root scripts

* remove cypress

* changed start command of frontend to match other scripts, use serve as webserver

* removed unnecessary test scripts

* cleaned up backend scripts

* fix install-app script for build-pipeline

* removed linting a second time in pipeline

* removing cypress dependencies

* Add section for beginners

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

sir-marc committed:
Merge remote-tracking branch 'origin/master' into improve-position-generation

sir-marc committed:
Fix linting errors

sir-marc committed:
Generate Units position on a free field

StuckiSimon commented on game-backend/src/helpers/gameObjects.ts:
```diff
@@ -11,3 +13,33 @@ export const countUnits = (state: State): number =>
 export const countFood = (state: State): number =>
   countGameObjectsPerType(state, GAME_OBJECT_TYPES.FOOD);
+
+export const generateFreePosition = (state: State): Position => {
```
Please add unit tests.

StuckiSimon commented on game-backend/src/helpers/gameObjects.ts:
```diff
@@ -11,3 +13,33 @@ export const countUnits = (state: State): number =>
+  );
+  const freeFields = fields.filter(
+    field =>
+      !!occupiedFields.find(
```
Watch out, this has bad runtime performance due to nested loop.
Why not just create a `map` of occupiedFields and access them by key (key e.g. `'${x}:${y}'`)?

StuckiSimon reviewed (changes requested):
Please create unit tests as part of the development phase.

sir-marc committed:
Refactor free fields search algorithm

sir-marc committed:
Add tests for free field search

michaelschufi committed:
AB#92 Backend Persistence Handling (#47)

* letting the client request its old session

* fixed based on feedback

* Extract ConnectionId to darwin-types

* fix tests

* added tests

* typing for matchUpdates

* cleaned up tests again

* renamed UserContextId to UserId and merged it with connectionId

* refactored server too

* deleted unnecessary type

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

StuckiSimon committed:
Implement Persistence Frontend Handling (#52)

mailaenderli committed:
#AB71 move to food (#48)

* implement food selector

* add food-selector tests

* implement userUnit-selector

* unify state-selector imports

* add userUnit-selector tests

* update help text

* Use pre html tag for CodeBlock.tsx

Preserves whitespace in code snippets.

* Improve HelpText

* improve coding style according to PR review

* Refactor gameObjects.ts helper

* move tests to gameObjects.test.ts

* improved tests based on PR feedback

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

StuckiSimon committed:
Use devicePixelRatio for canvas resolution (#53)

Makes canvas shapes appear sharper.

michaelschufi commented:
Looks good so far. While you're at it, could you make the MainController's `generateUnit` method use the new function?

sir-marc committed:
AB#83 food frontend rendering (#51)

* Add render layer order

* Wrap PixiGraphics component in custom component

* Add fill property to geometric graphics

* Refactor and redesign canvas

* Add triangle component

* Make Unit representation a bit fancier

* Add ownership differentiating styles

sir-marc committed:
Clean up after merge

sir-marc committed:
Merge branch 'master' into improve-position-generation

StuckiSimon commented on game-backend/src/helper/gameObjects.ts:
```diff
@@ -18,3 +23,41 @@ export const countUnits = (state: State): number =>
 export const countFood = (state: State): number =>
   countGameObjectsPerType(state, GAME_OBJECT_TYPES.FOOD);
+
+export const getFreeFields = (state: State) => {
```
Is there any correlation between `getGameObjectsPerType` and `getFreeFields`? I'd suggest creating a separate helper. Potentially changing the naming structure for these helpers would make sense.

StuckiSimon commented on game-backend/src/helper/gameObjects.ts:
```diff
@@ -18,3 +23,41 @@ export const countUnits = (state: State): number =>
+        ...acc,
+        ...new Array(ARENA_HEIGHT)
+          .fill(undefined)
+          .map((omit, j) => ({ x: i, y: j })),
```
Why do you use `_` and `omit` as variable names for omitted values? (In case it's just for handling eslint, you can easily configure this rule https://eslint.org/docs/rules/no-shadow#allow)

StuckiSimon replied:
Why do you use `_` and `omit` as variable names for omitted values? (In case it's just for handling eslint, you can easily configure this rule https://eslint.org/docs/rules/no-shadow#allow)

StuckiSimon committed:
Tune Canvas Element Styling (#54)

Change coloring and visualization of game objects.
Use inner shapes such as parallelogram.

mailaenderli commented on game-backend/src/helper/gameObjects.test.ts:
```diff
@@ -73,4 +75,22 @@ describe('getGameObjectsPerType', () => {
+      ])
+    );
+    expect(freeFields).toEqual(expect.arrayContaining([{ x: 1, y: 2 }]));
+  });
```
Shouldn't this test be in its own `describe {}` block because the tested function is `getFreeFields` not `getGameObjectsPerType`?

sir-marc committed:
Extract fields helper into separate file

StuckiSimon commented on game-backend/src/helper/fields.ts:
```diff
@@ -0,0 +1,42 @@
+import { ARENA_WIDTH, ARENA_HEIGHT } from '../../../darwin-types/Arena';
+
+export const getFreeFields = (state: State): Position[] => {
+  /* eslint no-shadow: ["error", { "allow": ["_"] }]*/
```
Why not add this to the global eslint config? Otherwise everyone has to configure this by himself.

sir-marc committed:
Handle move blocking (#50)

* Extend gameObject interface with moveBlocking prop

* Check moveBlocking prop in moveIntent

* AB#92 Backend Persistence Handling (#47)

* letting the client request its old session

* fixed based on feedback

* Extract ConnectionId to darwin-types

* fix tests

* added tests

* typing for matchUpdates

* cleaned up tests again

* renamed UserContextId to UserId and merged it with connectionId

* refactored server too

* deleted unnecessary type

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

* Implement Persistence Frontend Handling (#52)

* #AB71 move to food (#48)

* implement food selector

* add food-selector tests

* implement userUnit-selector

* unify state-selector imports

* add userUnit-selector tests

* update help text

* Use pre html tag for CodeBlock.tsx

Preserves whitespace in code snippets.

* Improve HelpText

* improve coding style according to PR review

* Refactor gameObjects.ts helper

* move tests to gameObjects.test.ts

* improved tests based on PR feedback

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

* Use devicePixelRatio for canvas resolution (#53)

Makes canvas shapes appear sharper.

* AB#83 food frontend rendering (#51)

* Add render layer order

* Wrap PixiGraphics component in custom component

* Add fill property to geometric graphics

* Refactor and redesign canvas

* Add triangle component

* Make Unit representation a bit fancier

* Add ownership differentiating styles

* Extract gameObject creation into helper function

* AB#92 Backend Persistence Handling (#47)

* letting the client request its old session

* fixed based on feedback

* Extract ConnectionId to darwin-types

* fix tests

* added tests

* typing for matchUpdates

* cleaned up tests again

* renamed UserContextId to UserId and merged it with connectionId

* refactored server too

* deleted unnecessary type

Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>

* Use devicePixelRatio for canvas resolution (#53)

Makes canvas shapes appear sharper.

* Tune Canvas Element Styling (#54)

Change coloring and visualization of game objects.
Use inner shapes such as parallelogram.

* Fix linting errors

* Fix unit creation helper

Co-authored-by: Michael Schaufelberger <michael.schufi@gmail.com>
Co-authored-by: Simon Stucki <simon_stucki@hotmail.com>
Co-authored-by: mailaenderli <61315599+mailaenderli@users.noreply.github.com>

sir-marc committed:
Merge branch 'master' into improve-position-generation

sir-marc committed:
Extract no-shadow eslint config into global config

StuckiSimon commented on game-backend/src/MainController.ts:
```diff
@@ -109,10 +110,8 @@ export default class MainController {
-        x: Math.floor(Math.random() * 20),
-        y: Math.floor(Math.random() * 20),
-      },
+      health: 100,
```
Please extract the default health either to a constant or don't pass it as it's already handled as part of `createUnit`.

StuckiSimon commented on game-backend/src/helper/fields.ts:
```diff
@@ -0,0 +1,41 @@
+      ],
+      []
+    );
+  const occupiedFields: Record<string, boolean> = state.objectIds.reduce(
```
What about extracting the steps (`getFlatFieldArray`, `getOccupiedFieldMap`) to separate functions? This makes the general logic easier to understand without knowing implementation details.

StuckiSimon reviewed (commented):
Some general notes

sir-marc committed:
Extract fields helper functions

sir-marc committed:
Remove unnecessary hard-coded health value in unitCreation

StuckiSimon reviewed (approved):
Looks good now 👍
